File: java/RomantoInteger.java
/*
 * Filename: RomantoInteger.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Wed 11 Mar 2015 03:41:17 PM EDT
 */

public class Solution {
	public int romanToInt(String s) {
		int ret = 0;

		for (int i = 0; i < s.length(); i++) {
			switch(s.charAt(i)) {
				case 'I':
					if (i < s.length() - 1) {
						switch(s.charAt(i+1)) {
							case 'V':
								ret += 4;
								i++;
								break;
							case 'X':
								ret += 9;
								i++;
								break;
							default:
								ret++;
								break;
						}
					} else {
						ret++;
					}
					break;
				case 'V':
					ret += 5;
					break;
				case 'X':
					if (i < s.length() - 1) {
						switch(s.charAt(i+1)) {
							case 'L':
								ret += 40;
								i++;
								break;
							case 'C':
								ret += 90;
								i++;
								break;
							default:
								ret += 10;
								break;
						}
					} else {
						ret += 10;
					}
					break;
				case 'L':
					ret += 50;
					break;
				case 'C':
					if (i < s.length() - 1) {
						switch(s.charAt(i+1)) {
							case 'D':
								ret += 400;
								i++;
								break;
							case 'M':
								ret += 900;
								i++;
								break;
							default:
								ret += 100;
								break;
						}
					} else {
						ret += 100;
					}
					break;
				case 'D':
					ret += 500;
					break;
				case 'M':
					ret += 1000;
					break;
				default:
					break;
			}
		}

		return ret;
	}
}

File: cpp/romanToInteger.cpp
/*
 * Filename: romanToInteger.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Tue 26 Jan 2016 08:48:09 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution
{
	public:
		int romanToInt(string s)
		{
			int res = 0;
			int i = 0;

			while (i < s.length())
			{
				switch (s[i])
				{
					case 'M':
						res += 1000;
						i++;
						break;
					case 'D':
						res += 500;
						i++;
						break;
					case 'C':
						if (i < s.length() - 1 && s[i+1] == 'M')
						{
							res += 900;
							i += 2;
							break;
						}
						else if (i < s.length() - 1 && s[i+1] == 'D')
						{
							res += 400;
							i += 2;
							break;
						}
						else
						{
							res += 100;
							i++;
							break;
						}
					case 'L':
						res += 50;
						i++;
						break;
					case 'X':
						if (i < s.length() - 1 && s[i+1] == 'C')
						{
							res += 90;
							i += 2;
							break;
						}
						else if (i < s.length() - 1 && s[i+1] == 'L')
						{
							res += 40;
							i += 2;
							break;
						}
						else
						{
							res += 10;
							i++;
							break;
						}
					case 'V':
						res += 5;
						i++;
						break;
					case 'I':
						if (i < s.length() - 1 && s[i+1] == 'X')
						{
							res += 9;
							i += 2;
							break;
						}
						else if (i < s.length() - 1 && s[i+1] == 'V')
						{
							res += 4;
							i += 2;
							break;
						}
						else
						{
							res++;
							i++;
							break;
						}
					default:
						break;
				}
			}

			return res;
		}
};
